### ISSUE TITLE
meson introspect doesn't handle expressions as possitional args to build targets correctly

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#5286]
currently Mesa cannot be introspected without a build directory, we end up trying to pop an empty list in build_targets.

That seemed easy enough to fix, but that just turns up other problems. It turns out that we don't correctly handle build targets that use expressions for their name, in mesa that looks like this:

```meson
shared_library('mylib' + suffix, ['sources'])
```
However, I've tested with string replacement and that has the same problem:
```meson
shared_library('mylib@0@'.format(suffix), 'one source')
```

modifying the 'unit/55 introspection' test with the following content will replicate:
```meson
# Using a an expresion for the name and a list for sources is intentional
t1 = executable('test' + '1', ['t1.cpp'], link_with: [sharedlib], install: true, build_by_default: get_option('test_opt2'))
t2 = executable('test@0@'.format('2'), 't2.cpp', link_with: [staticlib])
t3 = executable('test3', 't3.cpp', link_with: [sharedlib, staticlib], dependencies: [dep1])
```

I've tried to resolve this, but I'm not sure how the code is supposed to work.

Can be replicated with master and 0.50.1

**Comment by dcbaker** [comment]
@mensinda, I think this is your area?

**Comment by mensinda** [comment]
Yes, I will look into it.

**Comment by mensinda** [comment]
Technically, this isn't really a bug but a limitation of the current AST interpreter. There only a subset of a meson file is really interpreted. That being said, these cases should still be handled as best as possible. I am just not sure if this should go into 0.50.2 or 0.51.0, since fixing this is more like a new feature...

**Comment by mensinda** [comment]
Should be fixed by #5292
